platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,895 @@
+
+![TSS Platform Wiring](./TSS-Wiring-Diagram.drawio.png)
+
+#### RosterInitializer
+
+On software upgrade, the `RosterInitializer` is responsible for checking that there is enough key material for a
+candidate roster to be adopted as the active roster. If there is not enough key material, the candidate roster is
+not adopted and the active roster remains the active roster. If there is enough key material, the candidate roster
+is adopted as the active roster and the active roster becomes the previous roster.
+
+A roster rotation must be recorded in the state before the rest of the platform uses the state.
+
+#### TSS State Manager
+
+The TSS state manager is executed on the transaction handling thread and is capable of reading and writing the state
+during round handling. The logic will be encapsulated in a class with a well-defined API, but this will not be a
+full-fledged wiring component outside the consensus round handler.
+
+Responsibilities:
+
+1. Handle `TssMessage` system transactions
+    - handled in consensus order
+    - insert into the TSS data map if it is legal to do so
+        - don’t insert multiple messages for the same node
+        - don’t insert any messages if voting window has closed
+    - if inserted into the TSS data map, ensure that the message is forwarded to the TSS Message Validator
+    - if it is a self TSS message, forward the message to the TSS Message creator.
+2. Handle `TssVote` system transactions
+    - handled in consensus order
+    - insert into TSS data map if it is legal to do so
+        - don’t insert multiple votes for the same node
+        - don’t insert once voting is closed
+    - after insertion, re-tally votes and decide if voting is now closed
+    - if voting is closed as the result of a new vote,
+        - if the “no” votes win, then the candidate roster will never be adopted
+        - if the “yes” votes win, the candidate roster is now officially confirmed, and will be adopted at the next
+          upgrade boundary
+3. When a new candidate roster is set.
+    - inform the TSS message creator that there is a new candidate roster
+    - inform the TSS message validator that there is a new candidate roster
+4. Update Metrics
+    - indicate whether there is a candidate roster
+    - the current number of TSS messages collected for the candidate roster
+    - the current number of votes collected for the candidate roster
+        - keep different metrics both for “no” votes and “yes” votes
+5. On First Round After Restart (After PCES Replay and Reconnect)
+    - if there is no candidate roster, take no action
+    - if voting is closed, take no action
+    - if this node’s TSSMessages are not recorded in the TSSData map, send a message to the TSS message creator (so
+      it can resubmit the node’s TSS message)
+    - if this node’s vote is not recorded in the TSS data map, inform the TSS message validator that we still need to do
+      validation.
+
+#### TSS Message Creator
```

Comment:
Is it possible for a `TssMessageTransaction` to not reach consensus and get handled?  Suppose the node's ISP went down for half an hour.  The event becomes ancient.   Suppose half the shares have their TssMessages get lost due to events going ancient, then we will never be able to key the candidate roster.   We have to have guaranteed `TssMessage` sending.    That means re-sending the message after we do not hear it come back to us.  

Do you see an alternate way of guaranteeing `TssMessages` are sent and handled up until the point that voting is closed?

Reply:
Are you proposing some other logic hearbeat the candidate roster to the `TssMessageCreator` until the TssMessage from the share is recorded?

Reply:
> Is it possible for a TssMessageTransaction to not reach consensus and get handled

This shouldn't be possible. Tss messages shouldn't be considered pre-consensus

> We have to have guaranteed TssMessage sending

Agreed. I see 2 potential ways that a TssMessage could struggle to reach consensus:

1. An event containing a TssMessage goes stale, but the node doesn't go down
    - In this case, the `TransactionResubmitter` should continually attempt to resubmit, until the transaction successfully reaches consensus
2. A node crashes during the keying process, either not having created a TssMessage at all, or having submitted a TssMessage that ended up going stale
    - In this case, the node would behave the same way it would any other time it comes online during a keying process. The node would start, observe that the state doesn't contain the requisite self TssMessages, and then generate the messages.

Reply:
I've updated the proposal to extend the `Transaction Resubmitter` to handle the new `TssMessageTransaction` and `TssVoteTransaction`